### PR TITLE
MGMT-9914: Integration of the ODF LVM operator with assisted installer

### PR DESCRIPTION
**Pull request by dvachhan**
ODF LVMO should be available as a standalone (outside of Openshift Virtualization) deployment when a user installs a single-node openshift (SNO) cluster with assisted-installer

## List all the issues related to this PR

- [x] New Feature 
- [ ] Enhancement 
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [ ] Cloud
- [ ] Operator Managed Deployments
- [x] None

## How was this code tested?

- [x] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [ ] No tests needed

## Assignees

/cc @rewantsoni @avishayt @gobindadas @lalon4 

## Checklist

- [x] Title and description added to both, commit and PR.
- [x] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [x] Reviewers have been listed
- [x] This change does not require a documentation update (docstring, `docs`, README, etc)
- [x] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

**Review by rewantsoni — CHANGES_REQUESTED** [review]

**Comment by akalenyu** [comment]
Should we disable HPP entirely then?
AFAIU that is the decision that was made, and we want to avoid a situation where both are installed?
Knob should be available at https://github.com/openshift/assisted-service/blob/066bff08948a41b00785285b3e0879ecd8498e12/openshift/template.yaml#L114-L115

File: internal/operators/lvm/config.go [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+	"fmt"
+
+	"github.com/openshift/assisted-service/models"
+	"github.com/openshift/assisted-service/pkg/conversions"
+)
+
+type Config struct {
+	LvmCPUPerHost          int64  `envconfig:"LVM_CPU_PER_HOST" default:"1"`
+	LvmMemoryPerHostMiB    int64  `envconfig:"LVM_MEMORY_PER_HOST_MIB" default:"1200"`
+	LvmMinDiskSizeGB       int64  `envconfig:"LVM_MIN_DISK_SIZE_GB" default:"0"`
```
Comment (gamli75) [inline internal/operators/lvm/config.go:13]: Why 0?  and I don't see that we use this parameter

File: internal/operators/lvm/config.go [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+	LvmMinOpenshiftVersion string `envconfig:"LVM_MIN_OPENSHIFT_VERSION" default:"4.12.0"`
```
Comment (gamli75) [inline internal/operators/lvm/config.go:14]: We want to add LVMO in 4.11 as a tech preview -  why the default is 4.12?

File: internal/operators/lvm/lvm_operator.go [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+	ocpVersion, err = version.NewVersion(cluster.OpenshiftVersion)
+	if err != nil {
+		return api.ValidationResult{Status: api.Failure, ValidationId: o.GetHostValidationID(), Reasons: []string{err.Error()}}, nil
+	}
+	minOpenshiftVersionForLvm, err = version.NewVersion(o.config.LvmMinOpenshiftVersion)
+	if err != nil {
+		return api.ValidationResult{Status: api.Failure, ValidationId: o.GetHostValidationID(), Reasons: []string{err.Error()}}, nil
+	}
+	if ocpVersion.LessThan(minOpenshiftVersionForLvm) {
+		message := "ODF LVM operator is only supported for openshift versions 4.12.0 and above"
```
Comment (gamli75) [inline internal/operators/lvm/lvm_operator.go:91]: same remark as before - we want it as tech preview in 4.11

File: internal/operators/lvm/manifest.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+kind: OperatorGroup
+metadata:
+  name: openshift-storage-operatorgroup
+  namespace: "{{.OPERATOR_NAMESPACE}}"
+spec:
+  targetNamespaces:
+  - "{{.OPERATOR_NAMESPACE}}"`
+
+const LvmCluster = `apiVersion: lvm.topolvm.io/v1alpha1
+kind: LVMCluster
```
Comment (akalenyu) [inline internal/operators/lvm/manifest.go:106]: Are we not creating a `StorageClass` as well? does it get created by some operator?
Just thinking we might be better off creating our own one and setting it as default

**Review by akalenyu — COMMENTED** [review]
Looks good!

Mainly concerned with how HPP is disabled with this PR. I might be missing it entirely
@aglitke

**Comment by gamli75** [comment]
> Looks good!
> 
> Mainly concerned with how HPP is disabled with this PR. I might be missing it entirely (current IIUC they will both get deployed on SNO installs) cc @aglitke

We have a different epic for replacing HPP, see: https://issues.redhat.com/browse/MGMT-11046 
the requirements are:
image

Do you think that there will be a problem if the user will install CNV(HPP + LSO) & LVMO in OCP 4.11?

**Comment by akalenyu** [comment]
> > Looks good!
> > Mainly concerned with how HPP is disabled with this PR. I might be missing it entirely (current IIUC they will both get deployed on SNO installs) cc @aglitke
> 
> We have a different epic for replacing HPP, see: https://issues.redhat.com/browse/MGMT-11046 the requirements are: image
> 
> Do you think that there will be a problem if the user will install CNV(HPP + LSO) & LVMO in OCP 4.11?

Yes, the first thing that comes to mind is the extra non-installation disk requirement;
HPP also requires that so I imagine we'll have them fighting over this.

Another nuance is which storage class takes the "default" place.
4.11 would have HPP as default, and 4.12 will have LVMO as the default.
I think in both cases we'll have to create the storage class for LVMO ourselves to control the default storage class annotation,
WDYT?

**Comment by gamli75** [comment]
closed this PR - was replaced by https://github.com/openshift/assisted-service/pull/4197

**Pull request closed without merging**